realm/realm-library/src/main/cpp/util.cpp

```diff
@@ -43,195 +43,170 @@ jmethodID session_error_handler;
     catch (exception& e) {
-        ss << e.what() << " in " << file << " line " << line;
-        ThrowException(env, FatalError, ss.str());
+        std::ostringstream ss;
+        ss << "Unrecoverable error: " << e.what();
+        j_exception = JThrowableNew(env, realm_error, ss.str().c_str());
     }
-    /* catch (...) is not needed if we only throw exceptions derived from std::exception */
-}
 
-void ThrowException(JNIEnv* env, ExceptionKind exception, const char *classStr)
-{
-    ThrowException(env, exception, classStr, "");
-}
+    /* Catch (...) is not needed if we only throw exceptions derived from std::exception */
 
-void ThrowException(JNIEnv* env, ExceptionKind exception, const std::string& classStr, const std::string& itemStr)
-{
-    string message;
-    jclass jExceptionClass = NULL;
-
-    TR_ERR(env, "jni: ThrowingException %d, %s, %s.", exception, classStr.c_str(), itemStr.c_str())
-
-    switch (exception) {
-        case ClassNotFound:
-            jExceptionClass = env->FindClass("java/lang/ClassNotFoundException");
-            message = "Class '" + classStr + "' could not be located.";
-            break;
-
-        case IllegalArgument:
-            jExceptionClass = env->FindClass("java/lang/IllegalArgumentException");
-            message = "Illegal Argument: " + classStr;
-            break;
-
-        case IndexOutOfBounds:
-            jExceptionClass = env->FindClass("java/lang/ArrayIndexOutOfBoundsException");
-            message = classStr;
-            break;
-
-        case UnsupportedOperation:
-            jExceptionClass = env->FindClass("java/lang/UnsupportedOperationException");
-            message = classStr;
-            break;
-
-        case OutOfMemory:
-            jExceptionClass = env->FindClass("io/realm/internal/OutOfMemoryError");
-            message = classStr + " " + itemStr;
-            break;
-
-        case FatalError:
-            jExceptionClass = env->FindClass("io/realm/exceptions/RealmError");
-            message = "Unrecoverable error. " + classStr;
-            break;
-
-        case RuntimeError:
-            jExceptionClass = env->FindClass("java/lang/RuntimeException");
-            message = classStr;
-            break;
-
-        case BadVersion:
-            jExceptionClass = env->FindClass("io/realm/internal/async/BadVersionException");
-            message = classStr;
-            break;
-
-        case IllegalState:
-            jExceptionClass = env->FindClass("java/lang/IllegalStateException");
-            message = classStr;
-            break;
-
-        // Should never get here.
-        case ExceptionKindMax:
-        default:
-            break;
-    }
-    if (jExceptionClass != NULL) {
-        env->ThrowNew(jExceptionClass, message.c_str());
-        TR_ERR(env, "Exception has been throw: %s", message.c_str())
+    if (j_exception != nullptr) {
+        env->Throw(j_exception);
+        TR_ERR_NO_VA_ARG(env, "Exception has been throw")
     }
     else {
```

Comment:
yes, `j_exception` should not be `null`. If it is, the process should be terminated. `runtime_error` would be good enough.